src/udisksutabmonitor.c

```diff
@@ -0,0 +1,237 @@
+udisks_utab_monitor_class_init (UDisksUtabMonitorClass *klass)
+{
+  GObjectClass *gobject_class = (GObjectClass *) klass;
+
+  gobject_class->finalize    = udisks_utab_monitor_finalize;
+  gobject_class->constructed = udisks_utab_monitor_constructed;
+}
+
+static int
+fs_has_user_options (struct libmnt_fs *fs, void *user_data)
+{
+  return mnt_fs_get_user_options (fs) != NULL ? 1 : 0;
+}
+
+static void
+reload_utab_entries (UDisksUtabMonitor *monitor)
+{
+  udisks_utab_monitor_invalidate (monitor);
+  udisks_utab_monitor_ensure (monitor);
+}
+
+static gboolean
+utab_changed_event (GIOChannel *channel,
+                    GIOCondition cond,
+                    gpointer user_data)
+{
+  UDisksUtabMonitor *monitor = UDISKS_UTAB_MONITOR (user_data);
+  gboolean need_reload;
+  int r;
+
+  if (cond & ~G_IO_IN)
+    return TRUE;
+
+  need_reload = FALSE;
+
+  do
+  {
+    const char *filename;
+    r = mnt_monitor_next_change (monitor->mn, &filename, NULL);
+
+    if (r == 0)
+      need_reload = TRUE;
+  } while (r == 0);
+
+  if (need_reload)
+    reload_utab_entries (monitor);
+
+  return TRUE;
+}
+
+static void
+udisks_utab_monitor_constructed (GObject *object)
+{
+  UDisksUtabMonitor *monitor = UDISKS_UTAB_MONITOR (object);
+
+  monitor->mn = mnt_new_monitor();
+  mnt_monitor_enable_userspace (monitor->mn, TRUE, NULL);
+
+  monitor->utab_channel = g_io_channel_unix_new (mnt_monitor_get_fd (monitor->mn));
+  monitor->utab_watch_source = g_io_create_watch (monitor->utab_channel, G_IO_IN);
+  g_source_set_callback (monitor->utab_watch_source, (GSourceFunc) utab_changed_event, monitor, NULL);
+  g_source_attach (monitor->utab_watch_source, g_main_context_get_thread_default ());
+  g_source_unref (monitor->utab_watch_source);
+
+  if (G_OBJECT_CLASS (udisks_utab_monitor_parent_class)->constructed != NULL)
+    (*G_OBJECT_CLASS (udisks_utab_monitor_parent_class)->constructed) (object);
+}
+
+/**
+ * udisks_utab_monitor_new:
+ *
+ * Creates a new #UDisksUtabMonitor object.
+ *
+ * Signals are emitted in the <link
+ * linkend="g-main-context-push-thread-default">thread-default main
+ * loop</link> that this function is called from.
+ *
+ * Returns: A #UDisksUtabMonitor. Free with g_object_unref().
+ */
+UDisksUtabMonitor *
+udisks_utab_monitor_new (void)
+{
+  return UDISKS_UTAB_MONITOR (g_object_new (UDISKS_TYPE_UTAB_MONITOR, NULL));
+}
+
+/**
+ * udisks_utab_monitor_get_entries:
+ * @monitor: A #UDisksUtabMonitor.
+ *
+ * Gets all /run/mounts/utab entries
+ *
+ * Returns: (transfer full) (element-type UDisksUtabEntry): A list of #UDisksUtabEntry objects that must be freed with g_list_free() after each element has been freed with g_object_unref().
+ */
+GList *
+udisks_utab_monitor_get_entries (UDisksUtabMonitor  *monitor)
+{
+  GList *ret;
+
+  g_return_val_if_fail (UDISKS_IS_UTAB_MONITOR (monitor), NULL);
+
+  udisks_utab_monitor_ensure (monitor);
+
+  ret = g_list_copy (monitor->utab_entries);
+  g_list_foreach (ret, (GFunc) g_object_ref, NULL);
```

Comment:
There's also ``g_list_copy_deep()`` for exactly this. And also ``g_slist_copy_deep()``.